OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -39,6 +43,203 @@ static double UTOL = (double)SimTK::Eps*1e2;
+
+template <>
+struct std::hash<SmoothSegmentedFunctionParameters> final
+{
+    size_t operator()(const SmoothSegmentedFunctionParameters& params) const {
+        return HashOf(
+            params._mX, params._mY,
+            params._x0, params._x1,
+            params._y0, params._y1,
+            params._dydx0, params._dydx1,
+            params._computeIntegral,
+            params._intx0x1
+        );
+    }
+};
+
+//=============================================================================
+// SMOOTHSEGMENTEDFUNCTION DATA
+//=============================================================================
+
+namespace OpenSim {
+struct SmoothSegmentedFunctionData {
+        SmoothSegmentedFunctionData(
+                const SmoothSegmentedFunctionParameters& params,
+                const std::string& name);
+
+        /**Array of spline fit functions X(u) for each Bezier elbow*/
+        SimTK::Array_<SimTK::Spline> _arraySplineUX;
+        /**Spline fit of the integral of the curve y(x)*/
+        SimTK::Spline _splineYintX;
+
+        /**Bezier X1,...,Xn control point locations. Control points are
+        stored in 6x1 vectors in the order above*/
+        SimTK::Array_<SimTK::Vector> _mXVec;
+        /**Bezier Y1,...,Yn control point locations. Control points are
+        stored in 6x1 vectors in the order above*/
+        SimTK::Array_<SimTK::Vector> _mYVec;
+
+        /**The number of quintic Bezier curves that describe the relation*/
+        int _numBezierSections;
+};
+
+//=============================================================================
+// DATA LOOKUP
+//=============================================================================
+
+/** Manages an unordered map of SmoothSegmentedFunction's Data, using the Parameters as key.
+ * If the same SmoothSegmentedFunctionParameters were previously used to
+ * construct the SmoothSegmentedFunctionData, a shared pointer to that data is
+ * obtained. If the given parameters are new, a new data object is constructed
+ * and added. This prevents duplication of the SmoothSegmentedFunction data for
+ * identical curves.
+*/
+class SmoothSegmentedFunctionDataCache final {
+
+    public:
+    ~SmoothSegmentedFunctionDataCache() {}
+
+    std::shared_ptr<const SmoothSegmentedFunctionData> lookup(
+            const SmoothSegmentedFunctionParameters& params,
+            const std::string& name) {
+        std::lock_guard<std::mutex> guard{_cacheMutex};
+        auto it = _cache.find(params);
+        if (it != _cache.end()) {
+            return it->second;
+        } else {
+            std::shared_ptr<OpenSim::SmoothSegmentedFunctionData> ptr =
+                    std::make_shared<SmoothSegmentedFunctionData>(
+                            SmoothSegmentedFunctionData(params, name));
+            _cache[params] = ptr;
+            return ptr;
+        }
+    }
+
+    private:
+
+    std::mutex _cacheMutex;
+    std::unordered_map<SmoothSegmentedFunctionParameters,
```

Comment:
This cache strongly owns all function curves - even if no other system in OpenSim is using any. Combined with the use of a `static`-lifetime variable to hold this class, it's an unmanaged memory leak.

The `static` lifetime guarantees that the *cache* will only be destroyed at process-end, but strongly consider making the cache hold `std::weak_ptr<T>`'s rather than `std::shared_ptr<T>`'s, with the necessary `lock`ing, checking, reconstruction, etc. so that the curves are correctly freed when models are destroyed.